commerce/bookstore/python/requests.py

```diff
@@ -0,0 +1,254 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+from typedb.client import TypeDB, SessionType, TransactionType, TypeDBOptions
+from enum import Enum
+import argparse
+import config
+
+# Verbosity option implementation
+parser = argparse.ArgumentParser(description="Bookstore example requests")
+parser.add_argument("-v", "--verbose", "-d", "--debug", help="Increase output verbosity",
+                    dest="verbose", action="store_true")
+args = vars(parser.parse_args())
+
+if args["verbose"]:  # if the argument was set
+    print("High verbosity option turned on.")
+    debug = True  # Shows verbose debug messages in the console output
+else:
+    debug = False  # No debug messages
+
+
+class ResultCode(Enum):
+    OK = 0,
+    INPUT_INVALID = 1,
+    EXIT = 2
+
+
+def selection():  # This is the main UI to select a function to proceed with
+
+    print("Please choose one of the following functions: ")
+    print("1. Search for a book")
+    print("2. Search for a user")
+    print("3. Search for an order")
+    print("4. Search for books by genre")
+    print("0. Exit")
+    selection = input("Your request: ")  # Storing answer here
+    if selection == "":
+        print("Empty selection recognized. Please try again.")
+        return ResultCode.INPUT_INVALID
+    elif selection == "1":  # We chose variant #1 — searching for a book
+        search_book(input("Searching for a book. Please type in an ISBN or press enter for a full listing: "))
+        return ResultCode.OK
+    elif selection == "2":  # 2. Searching for a user
+        search_user(input("Searching for a user. Please type in a foreign ID or press enter for a full listing: "))
+        return ResultCode.OK
+    elif selection == "3":  # 3. Searching for an order
+        search_order(input("Searching for an order. Please type in an order ID or press enter for a full listing: "))
+        return ResultCode.OK
+    elif selection == "4":  # 4. Searching for books by genre
+        show_all_genres()  # Display all genres as a tip
+        search_genre(input("Searching for books by genre. Please type in genre name: "))
+        return ResultCode.OK
+    elif selection == "0" or "exit" or "exit()" or "close" or "close()" or "help":  # Exit the program
+        return ResultCode.EXIT
+    else:
+        print("Invalid selection recognized. Please try again.")  # Something else / unrecognized - repeat
+        return ResultCode.INPUT_INVALID
+
+
+def search_book(ISBN):  # Search book by ISBN (or show all books if empty ISBN given)
+
+    if ISBN == "":  # empty ISBN given
+        print("Empty input. Listing all books")
+        show_all_books()  # Display all books
+        return
+    else:  # Non-empty ISBN given
+        with TypeDB.core_client("localhost:1729") as client:
```

Comment:
We have a slight "anti-pattern" here (ie: malpractice): the `if` and `else` blocks are highly asymmetric.

The `if` block is super simple: "if ISBN field is empty, display all books." The "else" block says "if ISBN field is non-empty, open a TypeDB client, session and transaction, perform read query, iterate answers, open another client ..."

To fix this asymmetry, we should introduce a method that encapsulates all the behaviour of the `else` block - call it something like `show_single_book`, with parameter `ISBN`. (And we probably want to check again in this new method that ISBN is non-empty. But this time, if it's empty, throw an error. We can use `assert` for this.)

Apply the same change throughout this file.

Reply:
Fixed in https://github.com/vaticle/typedb-examples/pull/143/commits/822564ebd97ed3977de008aee6383cbf9f4587ab and https://github.com/vaticle/typedb-examples/pull/143#:~:text=822564e-,Finish%20refactoring%20functions,-%E2%80%A6